sync/sync.go

```diff
@@ -617,3 +628,121 @@ func (t *Task) getToReplica(address string) (rest.Replica, error) {
 
 	return rest.Replica{}, fmt.Errorf("Failed to find target replica to copy to")
 }
+
+// InternalSnapshotCleaner should be run in the background, it tries to delete a snapshot every 60s
+// It fetches the checkpoint from controller which is present at in-memory of controller.
+// If checkpoint is not available at controller, snapshot delete is not initiated.
+// A deletion candidate list is prepared
+// In each iteration of the loop, one snapshot is picked from the top and deleted.
+func (t *Task) InternalSnapshotCleaner(s *replica.Server, repClient *replicaClient.ReplicaClient) {
+	ticker := time.NewTicker(SnapshotDeletionInterval)
+
+	for range ticker.C {
+		snapshot, err := t.client.GetCheckpoint()
+		if err == nil && snapshot == s.Replica().Info().Checkpoint {
+			sortedSnapshotList, _ := getDeleteCandidateChain(s, snapshot)
+			if len(sortedSnapshotList) < 10 {
+				continue
+			}
+			var snap SnapList
+			for _, snap = range sortedSnapshotList {
+				if snap.name != "" {
+					break
+				}
+			}
+			if snap.name == "" {
+				continue
+			}
+			ops, err := s.PrepareRemoveDisk(snap.name)
+			if err != nil {
+				logrus.Errorf("PrepareRemoveDisk failed, err: %v", err)
+				continue
+			}
+			for _, op := range ops {
+				switch op.Action {
+				case replica.OpCoalesce:
+					logrus.Infof("Coalescing %v to %v", op.Target, op.Source)
+					if err = repClient.Coalesce(op.Source, op.Target); err != nil {
+						break
+					}
+				case replica.OpRemove:
+					logrus.Infof("Remove %v", op.Source)
+					if err = s.RemoveDiffDisk(op.Source); err != nil {
+						break
+					}
+				}
+				if err != nil {
+					logrus.Errorf("Snapshot deletion failed, err: %v", err)
+					break
+				}
+			}
+		}
+	}
+}
+
+func isHeadDisk(diskName string) bool {
+	if strings.HasPrefix(diskName, "volume-head-") && strings.HasSuffix(diskName, ".img") {
+		return true
+	}
+	return false
+}
+
+type SnapList struct {
+	name string
+	size int64
+}
+
+// getDeleteCandidateChain returns the chain of snapshots that can be deleted
+// Chain is sorted based on the snapshot size before returning
+// All the snapshots in the chain are returned except:
+// 1. Head snapshot
+// 2. Last snapshot, snapshot just below head
+// 3. Base snapshot
+// 4. User created snapshots not marked as removed
+func getDeleteCandidateChain(s *replica.Server, checkpoint string) ([]SnapList, error) {
```

Comment:
can we have unit tests just for this function? cases like
- empty checkpoint
- chain length of 3
- not existing checkpoint
- user created and not removed
- user created and removed
- varying size of snapshots to get the sorted one

Reply:
I guess - hitting these cases helps, and can give us few more scenarios.

Reply:
added these tests